### ISSUE TITLE
Implement `PopupButton` as replacement for wrongly used `MenuButton`s

### ISSUE THREAD

**Issue by Maran23** [godotengine/godot-proposals#6115]
### Describe the project you are working on

A game with heavy use of UI as well as the Godot editor.

### MenuButton

A `MenuButton` is flat and can not be focused by default.
This makes sense, as in all applications I know the menu can NOT be focused with the keyboard (or mouse), no matter if a native menu or not. To navigate to a `Menu`/`MenuButton`, mnemonic are typically used (e.g. holding ALT + F to navigate to **F**ile)
A menu entry is also typically flat in other applications too, so that is fine as well.

The problem is that we use `MenuButton`s not just for `Menu`s. We also use it when we need a `Button` (text or/with graphic) which shows a `Popup` on click.

Example in editor:

https://user-images.githubusercontent.com/66004280/213015016-865f74e4-3b72-4cca-8444-276a9e96f1bb.mp4

### Current Behaviour

The current behaviour results in some problems:
- Confusion as reported here: https://github.com/godotengine/godot/issues/70598 - 
->A `MenuButton` does not look like a `Button` with a dropdown. When used inside a `Menu` above, it looks obvious, but not outside of one.
- `MenuButton` can not be focused by the keyboard
- `MenuButton` style may not fit (as it is flat), although we use flat `Button`s a lot in the editor (but that might be different for games)

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
Example where a `MenuButton` fits right and where it is obvious that it will show a dropdown menu. 
We expect this behaviour as the `Menu` is right under the the Windows title bar, which is pretty normal for Windows users and commonly used across applications.
![image](https://user-images.githubusercontent.com/66004280/213025121-4632920d-7e7f-4fc2-90ff-5c53c117f94e.png)

Example where a `MenuButton` is used wrong. It is not clear from the look that this will actually show a dropdown menu. You also can not navigate to it by mouse or keyboard.
![image](https://user-images.githubusercontent.com/66004280/213026065-75f44544-cded-4302-8796-fc2aa45b03eb.png)

### OptionButton

Can we use the `OptionButton`? -> NO.
The `OptionButton` is a very specialized control and acts like a typical `ComboBox` or `DropDown`.
- It shows the current selected item (icon + text)
-> You can not set a text/icon to the `OptionButton` itself
- It can have only one selected item (radio items)
- It is also only made for radio items (we can not show something else without many hacking into the popup)

### Goal

A new component `PopupButton`, which is similar to the `MenuButton`, but does not contain any logic related to `Menu`s or `MenuButton`s. It is not flat by default and focusable by mouse and keyboard.

It contains some methods from `MenuButton`, which are generic and useful for `PopupButton` like classes.
`MenuButton` extends from `PopupButton`.

Up for discussion: `OptionButton` extends from `PopupButton`.
This will reduce more duplicated code.

### Non-Goal

- Refactor any `Button`/`Popup`/`MenuButton` or `OptionButton` related code.
-> The idea is to just introduce a new component to fit the usecase above, shift some methods from `MenuButton` (maybe also `OptionButton`) and replace the wrongly used `MenuButton`s in the next iteration.
- We do not want to mimic the `OptionButton`. 
->Therefore no specialized logic will be in the `PopupButton` class.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The following diagram shows my idea:
![abc](https://user-images.githubusercontent.com/66004280/213020962-f13d5805-3c75-4cad-b559-a6f82f785006.png)

`PopupButton` contains generic code from `MenuButton` and `OptionButton`. This can save us some duplicate code as well.

- A `MenuButton` is used in `Menu`s of any kind and contains specialized logic for `Menu` handling
- A `OptionButton` is used when a component is required that shows the current selected option and where only one selection is allowed.
- A `PopupButton` is used outside of `Menu`s where a component is required that fits with all the other components (can be navigated by mouse and keyboard, does not look flat by default) and shows a `Popup` which can be modified as needed.

---
`PopupButton` UI:
![image](https://user-images.githubusercontent.com/66004280/213021039-3dc50009-e4f9-48fa-a151-6bea48605311.png)

---
One thing up for discussion: The current `MenuButton`s have ony behaviour called `switch_on_hover`, which is enabled on some `MenuButton`s. Resulting in this behaviour:

https://user-images.githubusercontent.com/66004280/213024632-2aeca56c-5e4e-406d-951a-c074d2245f59.mp4

AFAIK, this is the only occurence where this feature is used. 
I don't think this behaviour should be inside `PopupButton`, as this is a feature which makes only sense for `Menu`s.
We can decide whether we want to replace this `MenuButton`s still to keep that behaviour or if it worth to drop and replace by the `PopupMenu`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

All 'wrongly' used `MenuButton`s needs to be adjusted with at least the following two lines of code in the editor and game code:
`menu_button.set_flat(false);`
`menu_button.set_focus_mode(FOCUS_ALL);`
That does not fix the issue, that a `MenuButton` does not look like a control with a `Popup`, So a graphic may need to be set as well.

The code duplication in the editor can not be solved otherwise.

### Is there a reason why this should be core and not an add-on in the asset library?

I think this should be in core as it affects the editor itself, will reduce code duplications and will help game developers if they need a `Button` with a customized `Popup` outside of a `Menu`.

**Comment by KoBeWi** [comment]
This was actually implemented once in https://github.com/godotengine/godot/pull/63358, but got reverted.

**Comment by zinnschlag** [comment]
>This makes sense, as in all applications I know the menu can NOT be focused with the keyboard (or mouse), no matter if a native menu or not. To navigate to a Menu/MenuButton, mnemonic are typically used (e.g. holding ALT + F to navigate to File)

Just for reference: The ALT-key navigation is not universal. It exists primarily on Windows. Not that common on Linux (though a some Linux desktops do support it). No idea about Mac. And you absolute can navigate a menu with the keyboard.

**Comment by Maran23** [comment]
> > This makes sense, as in all applications I know the menu can NOT be focused with the keyboard (or mouse), no matter if a native menu or not. To navigate to a Menu/MenuButton, mnemonic are typically used (e.g. holding ALT + F to navigate to File)
> 
> Just for reference: The ALT-key navigation is not universal. It exists primarily on Windows. Not that common on Linux (though a some Linux desktops do support it). No idea about Mac. And you absolute can navigate a menu with the keyboard.

Yes, with ALT + some key. But otherwise you can not navigate into the menu. I don't mean navigating inside the menu, that of course works, but that is not what my proposal is about.

**Comment by Maran23** [comment]
> This was actually implemented once in [godotengine/godot#63358](https://github.com/godotengine/godot/pull/63358), but got reverted.

Thank you for pointing that out. Interesting that @YuriSizov had the same idea.
Reading the comments in the PR, we may even combine this proposal with the idea from @YuriSizov in https://github.com/godotengine/godot-proposals/issues/5197. 
This actually answers one question above, whether we want to have `switch_on_hover` in `PopupButton` or not.

**Comment by YuriSizov** [comment]
> Thank you for pointing that out. Interesting that @YuriSizov had the same idea.

Well, there is a noticeable lack of a generic popup button, which was a problem for me since I was making one for the anchor presets and container sizing. 🙃 While I did end up making it, the biggest missing piece was switch on hover, because of the way it was implemented. Making a base class to share that functionality seemed sensible, but turned quite hacky when trying to type the API, where extending classes need to return more specific versions of the Popup.

So that's how I ended up with my proposal, where the biggest missing piece is shared between all buttons, and a potential PopupButton doesn't need to be related to MenuButton or ColorPickerButton.
